src/Algebra/Linear/Structures.agda

```diff
@@ -0,0 +1,58 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some linear algebraic structures.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Algebra.Linear.Structures where
+
+open import Algebra        using (CommutativeRing)
+open import Algebra.Module using (Module)
+open import Data.List      using (List; foldr)
+open import Data.Product
+open import Function
+open import Level          using (Level; _⊔_; suc)
+
+module _
+  {r ℓr m ℓm : Level}
+  {ring      : CommutativeRing r ℓr}
+  (mod       : Module ring m ℓm)
+  where
+
+  open CommutativeRing ring renaming (Carrier  to S)
+  open Module          mod  renaming (Carrierᴹ to V)
+
+  -- A set of "vectors" forms a basis for a space iff it is complete
+  -- under some inner product.
+  record BasisFor (_∙_ : V → V → S) : Set (suc (ℓm ⊔ m ⊔ r)) where
+
+    -- Scale a vector according to some reduction function.
+    vscale : (V → S) → V → V
+    vscale f = uncurry _*ₗ_ ∘ < f , id >
+
+    -- Generate a vector from the basis, according to some reduction function.
+    vgen : (V → S) → List V → V
+    vgen f = foldr (_+ᴹ_ ∘ vscale f) 0ᴹ
```

Comment:
Hmm the type of this function looks a bit weird as is, I assume the `List V` that it takes *is* the basis set declared below? Why not move the definition below the fields, and then pass `basisSet` to it explicitly?

Reply:
Because the `basisComplete` field calls it.